tests/test_consumer.py

```diff
@@ -2052,3 +2052,199 @@ async def on_partitions_assigned(self, assigned):
+        # and consumer2
+        await consumer2.start()
+        self.add_cleanup(consumer2.stop)
+        await consumer2.seek_to_committed()
+
+        assert consumer1._group_instance_id == "consumer_1_static_group_member"
+        assert consumer2._group_instance_id == "consumer_2_static_group_member"
+
+        c1_partitions = consumer1.assignment()
+        c2_partitions = consumer2.assignment()
+        # take note of assigned partitions
+        if c1_partitions == {tp0}:
+            c1_assignment = {tp0}
+            c2_assignment = {tp1}
+        else:
+            c1_assignment = {tp1}
+            c2_assignment = {tp0}
+
+        listener1.revoke_mock.assert_called_with({tp0, tp1})
+        self.assertEqual(listener1.revoke_mock.call_count, 2)
+        listener1.assign_mock.assert_called_with(c1_assignment)
+        self.assertEqual(listener1.assign_mock.call_count, 2)
+        listener2.revoke_mock.assert_called_with(set())
+        self.assertEqual(listener2.revoke_mock.call_count, 1)
+        listener2.assign_mock.assert_called_with(c2_assignment)
+        self.assertEqual(listener2.assign_mock.call_count, 1)
+
+        # confirm diff partitions
+        assert c2_partitions != c1_partitions
+        # unsubscribe from topic, check if
+        # a third rebalance occurs for consumer1.
+        # It should not since KIP-345 is active.
+        consumer2.unsubscribe()
+        self.assertEqual(listener1.revoke_mock.call_count, 2)
+        self.assertEqual(listener1.assign_mock.call_count, 2)
+        # ensure that consumer2's assigned partitions
+        # are not re-assigned to consumer1
+        for p in c2_partitions:
+            assert p not in c1_partitions
+        # reassign, ensure that it goes back to same partition
+        consumer2.subscribe(topics=[self.topic], listener=listener2)
+        await consumer2._subscription.wait_for_assignment()
+        # since consumer2 rejoins the group, it receives it's old
+        # assigned partitions. We assert that a rebalance for all
+        # partitions was not triggered by checking that the listener2
+        # call count is 2, which is the initial rebalance that assigned the
+        # inital partitions, and this assignment that reassigns the partitions
+        self.assertEqual(listener2.revoke_mock.call_count, 2)
+        self.assertEqual(listener2.assign_mock.call_count, 2)
+        # we check the other consumer's listener to make sure
+        # it did not undergo a rebalance. The call count should still
+        # be 2.
+        self.assertEqual(listener1.revoke_mock.call_count, 2)
+        self.assertEqual(listener1.assign_mock.call_count, 2)
+        assert c2_partitions == consumer2.assignment()
+
+        # wait for timeout, consumer1 should get all
+        # partitions after rebalance
+        all_partitions = frozenset(list(c1_partitions) + list(c2_partitions))
+        await consumer2.stop()
+        await asyncio.sleep(25)
```

Comment:
Could we document here exactly the reason this is 25s? As I understand it is derived from default timeouts, maybe we can make it as low as 10s?
The reason I am asking is that CI has limited time to execute and stops like those increase runtime constantly. A better approach would be to poll for results, for example, wait for every second till listener1.revoke_mock.call_count > 2.

Reply:
I'll make the polling change as recommended.